torchtext/prototype/datasets/text_classification.py

```diff
@@ -0,0 +1,628 @@
+        if fname.endswith('train.csv') and 'train' in data_select:
+            path['train'] = fname
+        if fname.endswith('test.csv') and 'test' in data_select:
+            path['test'] = fname
+
+    iters_group = {}
+    if 'train' in data_select:
+        iters_group['vocab'] = _csv_iterator(path['train'], tokenizer, ngrams)
+    for item in data_select:
+        iters_group[item] = _csv_iterator(path[item], tokenizer,
+                                          ngrams, yield_cls=True)
+    return iters_group
+
+
+def _generate_imdb_data_iterators(dataset_name, root, ngrams, tokenizer, data_select):
+    if not tokenizer:
+        tokenizer = get_tokenizer("basic_english")
+
+    if not set(data_select).issubset(set(('train', 'test'))):
+        raise TypeError('Given data selection {} is not supported!'.format(data_select))
+
+    dataset_tar = download_from_url(URLS[dataset_name], root=root)
+    extracted_files = extract_archive(dataset_tar)
+
+    iters_group = {}
+    if 'train' in data_select:
+        iters_group['vocab'] = _imdb_iterator('train', extracted_files, tokenizer, ngrams)
+    for item in data_select:
+        iters_group[item] = _imdb_iterator(item, extracted_files,
+                                           tokenizer, ngrams, yield_cls=True)
+    return iters_group
+
+
+def _setup_datasets(iters_group, vocab, removed_tokens, data_select):
+
+    if vocab is None:
+        if 'vocab' not in iters_group.keys():
+            raise TypeError("Must pass a vocab if train is not selected.")
+        logging.info('Building Vocab based on train data')
+        vocab = build_vocab_from_iterator(iters_group['vocab'])
+    else:
+        if not isinstance(vocab, Vocab):
+            raise TypeError("Passed vocabulary is not of type Vocab")
+    logging.info('Vocab has {} entries'.format(len(vocab)))
+
+    data = {}
+    for item in iters_group.keys():
+        data[item] = {}
+        data[item]['data'] = []
+        data[item]['labels'] = []
+        logging.info('Creating {} data'.format(item))
+        data_iter = _create_data_from_iterator(vocab, iters_group[item], removed_tokens)
+        for cls, tokens in data_iter:
+            data[item]['data'].append((torch.tensor(cls),
+                                       torch.tensor([token_id for token_id in tokens])))
+            data[item]['labels'].append(cls)
+        data[item]['labels'] = set(data[item]['labels'])
+
+    return tuple(TextClassificationDataset(vocab, data[item]['data'],
+                                           data[item]['labels']) for item in data_select)
+
+
+def _initiate_datasets(dataset_name, root='.data', ngrams=1, vocab=None,
+                       removed_tokens=[], tokenizer=None,
+                       data_select=('train', 'test')):
+    if isinstance(data_select, str):
+        data_select = [data_select]
+    if dataset_name == 'IMDB':
+        return _setup_datasets(_generate_imdb_data_iterators(dataset_name, root, ngrams,
```

Comment:
The only difference here is _generate_imdb_data_iterators and _generate_data_iterators. If we create a more generic version of these two functions that can accept an iters_group we should be able to write less code.

Reply:
fixed.